File: ShamirSecretSharing.java
package ShamirSecretSharing;

import java.math.BigInteger;
import java.util.*;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public class ShamirSecretSharing {

    public static class Point {
        public BigInteger x;
        public BigInteger y;

        public Point(BigInteger x, BigInteger y) {
            this.x = x;
            this.y = y;
        }
    }

    // Convert a number from given base to decimal using BigInteger
    public static BigInteger convertFromBase(String value, int base) {
        BigInteger result = BigInteger.ZERO;
        BigInteger baseValue = BigInteger.valueOf(base);
        BigInteger power = BigInteger.ONE;

        // Process digits from right to left
        for (int i = value.length() - 1; i >= 0; i--) {
            char digit = value.charAt(i);
            int digitValue;

            if (digit >= '0' && digit <= '9') {
                digitValue = digit - '0';
            } else if (digit >= 'a' && digit <= 'z') {
                digitValue = digit - 'a' + 10;
            } else if (digit >= 'A' && digit <= 'Z') {
                digitValue = digit - 'A' + 10;
            } else {
                throw new IllegalArgumentException("Invalid digit: " + digit);
            }

            if (digitValue >= base) {
                throw new IllegalArgumentException("Digit " + digit + " is invalid for base " + base);
            }

            result = result.add(BigInteger.valueOf(digitValue).multiply(power));
            power = power.multiply(baseValue);
        }

        return result;
    }

    // Lagrange interpolation to find the constant term (secret)
    public static BigInteger lagrangeInterpolation(List<Point> points) {
        BigInteger result = BigInteger.ZERO;
        int n = points.size();

        for (int i = 0; i < n; i++) {
            Point point_i = points.get(i);
            BigInteger xi = point_i.x;
            BigInteger yi = point_i.y;

            // Calculate Lagrange basis polynomial Li(0)
            BigInteger numerator = BigInteger.ONE;
            BigInteger denominator = BigInteger.ONE;

            for (int j = 0; j < n; j++) {
                if (i != j) {
                    Point point_j = points.get(j);
                    BigInteger xj = point_j.x;

                    // For Li(0): numerator *= (0 - xj) = -xj
                    numerator = numerator.multiply(xj.negate());
                    // denominator *= (xi - xj)
                    denominator = denominator.multiply(xi.subtract(xj));
                }
            }

            // Calculate yi * Li(0) = yi * (numerator / denominator)
            // Since we need exact division, we'll use BigInteger division
            BigInteger term = yi.multiply(numerator).divide(denominator);
            result = result.add(term);
        }

        return result;
    }

    public static void main(String[] args) {
        try {
            // Example with test case 1
            String jsonInput1 = """
            {
                "keys": {
                    "n": 4,
                    "k": 3
                },
                "1": {
                    "base": "10",
                    "value": "4"
                },
                "2": {
                    "base": "2",
                    "value": "111"
                },
                "3": {
                    "base": "10",
                    "value": "12"
                },
                "6": {
                    "base": "4",
                    "value": "213"
                }
            }""";

            ObjectMapper mapper = new ObjectMapper();
            JsonNode root = mapper.readTree(jsonInput1);

            int n = root.get("keys").get("n").asInt();
            int k = root.get("keys").get("k").asInt();

            List<Point> points = new ArrayList<>();

            // Parse points from JSON
            Iterator<String> fieldNames = root.fieldNames();
            while (fieldNames.hasNext()) {
                String fieldName = fieldNames.next();
                if (!fieldName.equals("keys")) {
                    JsonNode pointNode = root.get(fieldName);
                    int x = Integer.parseInt(fieldName);
                    int base = pointNode.get("base").asInt();
                    String value = pointNode.get("value").asText();

                    BigInteger y = convertFromBase(value, base);
                    points.add(new Point(BigInteger.valueOf(x), y));
                }
            }

            // Sort points by x coordinate
            points.sort((p1, p2) -> p1.x.compareTo(p2.x));

            // Take first k points for reconstruction
            List<Point> selectedPoints = points.subList(0, k);

            // Find the secret using Lagrange interpolation
            BigInteger secret = lagrangeInterpolation(selectedPoints);

            System.out.println("Secret: " + secret);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: ShamirSecretSharing.js

// Shamir's Secret Sharing Solution in JavaScript
// Run with: node ShamirSecretSharing.js

class BigNumber {
    constructor(value) {
        if (typeof value === 'string') {
            this.value = BigInt(value);
        } else if (typeof value === 'number') {
            this.value = BigInt(value);
        } else if (typeof value === 'bigint') {
            this.value = value;
        } else {
            this.value = BigInt(0);
        }
    }

    add(other) {
        return new BigNumber(this.value + other.value);
    }

    subtract(other) {
        return new BigNumber(this.value - other.value);
    }

    multiply(other) {
        return new BigNumber(this.value * other.value);
    }

    divide(other) {
        return new BigNumber(this.value / other.value);
    }

    toString() {
        return this.value.toString();
    }

    toNumber() {
        return Number(this.value);
    }
}

function convertFromBase(value, base) {
    let result = new BigNumber(0);
    let power = new BigNumber(1);
    const baseValue = new BigNumber(base);

    for (let i = value.length - 1; i >= 0; i--) {
        const digit = value[i];
        let digitValue;

        if (digit >= '0' && digit <= '9') {
            digitValue = digit.charCodeAt(0) - '0'.charCodeAt(0);
        } else if (digit >= 'a' && digit <= 'z') {
            digitValue = digit.charCodeAt(0) - 'a'.charCodeAt(0) + 10;
        } else if (digit >= 'A' && digit <= 'Z') {
            digitValue = digit.charCodeAt(0) - 'A'.charCodeAt(0) + 10;
        } else {
            throw new Error(`Invalid digit: ${digit}`);
        }

        if (digitValue >= base) {
            throw new Error(`Digit ${digit} is invalid for base ${base}`);
        }

        result = result.add(new BigNumber(digitValue).multiply(power));
        power = power.multiply(baseValue);
    }

    return result;
}

function lagrangeInterpolation(points) {
    let result = 0;
    const n = points.length;

    for (let i = 0; i < n; i++) {
        const xi = points[i].x;
        const yi = points[i].y;

        let li = 1;
        for (let j = 0; j < n; j++) {
            if (i !== j) {
                const xj = points[j].x;
                li *= (0 - xj) / (xi - xj);
            }
        }

        result += yi * li;
    }

    return Math.round(result);
}

function solveShamirSecretSharing(testCase) {
    const { keys, ...pointsData } = testCase;
    const { n, k } = keys;

    console.log(`n = ${n}, k = ${k}`);
    console.log(`Need ${k} points to reconstruct polynomial of degree ${k-1}`);

    const points = [];

    // Parse points from test case
    for (const [key, pointData] of Object.entries(pointsData)) {
        const x = parseInt(key);
        const base = parseInt(pointData.base);
        const value = pointData.value;

        const y = convertFromBase(value, base);
        points.push({
            x: x,
            y: y.toNumber(),
            originalBase: base,
            originalValue: value
        });

        console.log(`Point ${key}: x=${x}, y=${y.toString()} (from base ${base} value '${value}')`);
    }

    // Sort points by x coordinate
    points.sort((a, b) => a.x - b.x);

    // Take first k points for reconstruction
    const selectedPoints = points.slice(0, k);
    console.log(`\nSelected ${selectedPoints.length} points for reconstruction:`);
    selectedPoints.forEach(p => console.log(`(${p.x}, ${p.y})`));

    // Find the secret using Lagrange interpolation
    const secret = lagrangeInterpolation(selectedPoints);

    console.log(`\nSecret (constant term): ${secret}`);

    return secret;
}

// Test case 1
const testCase1 = {
    "keys": {
        "n": 4,
        "k": 3
    },
    "1": {
        "base": "10",
        "value": "4"
    },
    "2": {
        "base": "2", 
        "value": "111"
    },
    "3": {
        "base": "10",
        "value": "12"
    },
    "6": {
        "base": "4",
        "value": "213"
    }
};

// Test case 2
const testCase2 = {
    "keys": {
        "n": 10,
        "k": 7
    },
    "1": {
        "base": "6",
        "value": "13444211440455345511"
    },
    "2": {
        "base": "15",
        "value": "aed7015a346d635"
    },
    "3": {
        "base": "15", 
        "value": "6aeeb69631c227c"
    },
    "4": {
        "base": "16",
        "value": "e1b5e05623d881f"
    },
    "5": {
        "base": "8",
        "value": "316034514573652620673"
    },
    "6": {
        "base": "3",
        "value": "2122212201122002221120200210011020220200"
    },
    "7": {
        "base": "3",
        "value": "20120221122211000100210021102001201112121"
    },
    "8": {
        "base": "6",
        "value": "20220554335330240002224253"
    },
    "9": {
        "base": "12",
        "value": "45153788322a1255483"
    },
    "10": {
        "base": "7",
        "value": "1101613130313526312514143"
    }
};

console.log("=== Test Case 1 ===");
const secret1 = solveShamirSecretSharing(testCase1);

console.log("\n=== Test Case 2 ===");
const secret2 = solveShamirSecretSharing(testCase2);

console.log("\n=== Results ===");
console.log(`Test Case 1 Secret: ${secret1}`);
console.log(`Test Case 2 Secret: ${secret2}`);
